ggml/src/lib.rs

```diff
@@ -303,6 +303,23 @@ impl Context {
     pub fn used_mem(&self) -> usize {
         unsafe { ggml_sys::ggml_used_mem(self.ptr.as_ptr()) }
     }
+
+    /// Set scratch buffer
+    pub fn use_scratch(&self, scratch_buffer: Option<&Buffer>) {
```

Comment:
Shouldn't this be a `Option<&mut Buffer>`?

Reply:
the infered lifetime is wrong. The buffer is consumed.

`&mut Buffer` should at least live until `&self` is gone

Reply:
Signature could be

`pub fn use_scratch(&self, scratch_buffer: Option<&mut [u8]>)`

then `Buffer` isn't needed

Reply:
I have a feeling Rust's aliasing rules would not like this.

The context should have a field like
`_scratch: Option<PhantomData<&mut [u8]>>`

Reply:
Yes, definitely should be mut. I'll fix it.

And yes, it could be `Option<&mut [u8]>` but I don't think there's a safe way to create uninitialized `[u8; BUF_SIZE]`, so the `Buffer` helper at least simplifies that bit slightly. Or alternatively could just create `Vec<u8>` with wanted amount of `0`s, but that does of course cause "unneeded" runtime cost.

Reply:
Actually I also wonder if I should just make this function unsafe as well... It does crash to segmentation fault if it's used incorrecly. On the other hand I guess quite a lot of other functions end up being unsafe as well if that would be the reasoning.

Reply:
It being unsafe seems reasonable to me - if the caller has to make guarantees that aren't handled by Rust in order to use it safely, it should be `unsafe`